webapp/src/js/services/telemetry.js

```diff
@@ -0,0 +1,215 @@
+var moment = require('moment'),
+  uuid = require('uuid/v4');
+
+angular
+  .module('inboxServices')
+  .factory('Telemetry', function($log, $q, $window, DB, Session) {
+    'use strict';
+    'ngInject';
+
+    var queue = $q.resolve();
+
+    // Intentionally scoped to the whole browser (for this domain). We can then
+    // tell if multiple users use the same device
+    var DEVICE_ID_KEY = 'medic-telemetry-device-id';
```

Comment:
I'm not sure if the current implementation of
```
        uniqueDeviceId = uuid();
        $window.localStorage.setItem(DEVICE_ID_KEY, uniqueDeviceId);
```
will actually be shared by multiple users using the same device in practice.  Mobile users will "log out" to change user by clearing the device cache, which will presumably clear this `localStorage` value.  On desktop, if I use Chrome and then Firefox, the device ID would be different for each of them - despite being the same device.  

There are definitely more reliable ways we could fingerprint an Android device, e.g. by requesting info from Java.  Alternatively, renaming the "device id" to something more accurately reflecting its limitations could help communicate what this actually represents.

Reply:
> Mobile users will "log out" to change user by clearing the device cache, which will presumably clear this localStorage value. On desktop, if I use Chrome and then Firefox, the device ID would be different for each of them - despite being the same device.

Those two scenarios (clearing your cache completely, and using different browsers) will not be caught here no. However, if you just log out of your session and log back in as someone else using the same browser, e.g. at a facility or something, that would be caught.

We could use something from android, but then it wouldn't work outside of that, and ideally it would be consistent in shape (ie both should be UUIDs).

The core "real" problem we're solving here, is that if you are logged in on two different devices / browsers, then I don't want your aggregate telemetry documents to conflict with each other. It is icing on the cake if we could tie different users back to the same device / browser.

Do you think `sessionId` sounds more correct?

Reply:
> However, if you just log out of your session and log back in as someone else using the same browser, e.g. at a facility or something, that would be caught.

While this is an interesting scenario, perhaps not particularly interesting at least for load timings.  On the other hand, we currently have single-user-multi-login being used in prod projects with large datasets, where comparing performance and usage metrics for the same user, on the same device, but across his different accounts could be very interesting.  These users are very likely to have sllow startup times.

> We could use something from android, but then it wouldn't work outside of that, and ideally it would be consistent in shape (ie both should be UUIDs).

For metrics used for measuring performance, having a quick way to differentiate between android users and desktop users may be _very_ useful.  Also being able to differentiate between different apps on the same phone would be nice, for the scenario above.  Filtering by a combination of browser UserAgent and device ID would allow this.

My understanding is that getting a specific user ID for a desktop user isn't directly possible, but I don't think this means we should avoid doing it for mobile users.

> The core "real" problem we're solving here, is that if you are logged in on two different devices / browsers, then I don't want your aggregate telemetry documents to conflict with each other.
>
> Do you think sessionId sounds more correct?

If that's the aim, then using a random UUID sounds fine.  `sessionId` seems reasonable, as long as it's not already used elsewhere to mean something different.  I can only find one reference in the webapp repo, where `medic-api` refers to the `AuthSession` as `sessionId`, but this is an isolated case, and likely misleading in itself.